### ISSUE TITLE
Private & Public DHTs

### ISSUE THREAD

**Issue by Stebalien** [libp2p/go-libp2p#803]
Followup from a discussion with @aschmahmann.

Proposal: Run a second private DHT, mounted on `/lan/dht/proto`, in parallel to the public DHT.

With the next release of the DHT, DHT nodes will default to an "auto" mode that will only turn them into DHT servers when they learn that they are globally reachable. Unfortunately, as described in https://github.com/ipfs/go-ipfs/issues/6737, this will break offline/vpn/etc. DHTs.

The proposal is to run an additional DHT, on a separate protocol, with a special routing table filter that _only_ admits nodes we believe are on our local network (e.g., nodes where the remote address is non-routable, nodes with the same public IP, etc.). All (or maybe just auto?) nodes with DHT support will act as servers in this local DHT and these two DHTs will be queried in parallel.

This will help in three ways:

1. IPNS, content routing, etc. will transparently work in offline, disconnected networks and VPNs.
2. Content routing will be _very_ fast between peers in a local network.
3. Given proper address [_scopes_](https://github.com/libp2p/go-libp2p/issues/793), we can advertise local address records to this local DHT, and public address records to the global DHT, finally fixing https://github.com/libp2p/go-libp2p/issues/436.

**Comment by Stebalien** [comment]
In terms of implementation, this should be _very_ simple as we already have all the components. We need:

1. The ability to combine multiple DHTs: https://github.com/libp2p/go-libp2p-routing-helpers.
2. Custom protocol prefixes (already supported).
3. The ability to filter peers before we consider entering them into the routing table. We already have rudimentary support for this in the stabilize branch, we just need to be able to pass an arbitrary function into the constructor.

**Comment by bonedaddy** [comment]
Is this going to be heavily tested and benchmarked before being implemented? This seems like a massive change that will probably break functionality for more than one system in ways that are unknown to the relatively low test coverage across all repositories.

**Comment by Stebalien** [comment]
> Is this going to be heavily tested and benchmarked before being implemented?

Yes.

> This seems like a massive change that will probably break functionality for more than one system in ways that are unknown due to the relatively low test coverage across all repositories.

Well, we don't want undialable nodes to join the main DHT so, by default, nodes _won't_ join as servers unless they can determine that they're publicly dialable.

However, we _also_ don't want to break VPNs, offline networks, test networks, etc. Adding a second private DHT is the solution that's least likely to break those. In terms of overhead, "private" DHTs should be tiny in most cases so queries should take no time (in most cases, all nodes will be directly connected anyways).

**Comment by Stebalien** [comment]
Related: https://github.com/libp2p/go-libp2p-kad-dht/issues/564

**Comment by danrobi11** [comment]
Hello plz consider supporting Lokinet. Thx!
https://github.com/ipfs/go-ipfs/issues/7113

**Comment by willscott** [comment]
@danrobi11 - not related to this issue. please stay on topic

**Comment by willscott** [comment]
a lan/wan split DHT implementation is now merged into go-libp2p-kad-dht in https://github.com/libp2p/go-libp2p-kad-dht/pull/570

**Comment by Stebalien** [comment]
This has been released.
